File: frontend/src/service/usuarioService.ts
import { api } from './apiBase';
import type { ApiResponse, Page } from '../type/apiTypes';
import type {
    EditarPerfilRequestDTO,
} from '../type/requestTypes';
import type { Solicitud, Usuario } from '../type/entityTypes';
import { TipoCrypto, TipoSolicitud, TipoWallets } from '../type/enum';

// Response types
export interface JwtResponse {
    token: string;
    type: string;
    id: number;
    username: string;
    email: string;
    roles: string[];
    user: Usuario;
    usuarioEnRed: number;
}

export interface UsuarioEnRedResponse {
    id: number;
    username: string;
    email: string;
    nombre: string;
    apellido: string;
    fechaRegistro: string;
    nivel: number;
}

const BASE_PATH = '/usuarios';

// Editar perfil del usuario autenticado
// PUT /api/usuarios/perfil
const editarPerfil = async (editarPerfilRequest: EditarPerfilRequestDTO): Promise<ApiResponse<Usuario>> => {
    return api.put<Usuario>(`${BASE_PATH}/perfil`, editarPerfilRequest);
};

// Verificación de dos pasos
// POST /api/usuarios/verificacion-dos-pasos
const verificacionDosPasos = async (codigoVerificacion: string): Promise<ApiResponse<string>> => {
    return api.post<string>(`${BASE_PATH}/verificacion-dos-pasos`, null, {
        params: { codigoVerificacion }
    });
};

// Obtener usuarios en red
// GET /api/usuarios/red/{username}
const obtenerUsuariosEnRed = async (username: string): Promise<ApiResponse<UsuarioEnRedResponse[]>> => {
    return api.get<UsuarioEnRedResponse[]>(`${BASE_PATH}/red/${username}`);
};

// Editar usuario (Admin)
// PUT /api/usuarios/admin/editar
const editarUsuarioAdmin = async (usuario: Usuario): Promise<ApiResponse<string>> => {
    return api.put<string>(`${BASE_PATH}/admin/editar`, usuario);
};

// Solicitar compra de licencia
// POST /api/usuarios/solicitar-licencia
const solicitarCompraLicencia = async (tipoCrypto: TipoCrypto, tipoLicencia: string, tipoSolicitud: TipoSolicitud): Promise<ApiResponse<string>> => {
    return api.post<string>(`${BASE_PATH}/solicitar-licencia`, null, {
        params: {
            tipoCrypto: TipoCrypto[tipoCrypto], // Convertir enum a string
            tipoLicencia,
            tipoSolicitud: TipoSolicitud[tipoSolicitud] // Convertir enum a string
        }
    });
};

//Obtener todas las solicitudes PENDIENTES (Admin)
// GET /api/usuarios/admin/solicitudes
const obtenerSolicitudes = async (page: number = 0, size: number = 10, sort?: string): Promise<ApiResponse<Page<Solicitud>>> => {
    return api.get<Page<Solicitud>>(`${BASE_PATH}/admin/solicitudes`, {
        params: {
            page,
            size,
            ...(sort && { sort })
        }
    });
};

// Obtener todos los usuarios con filtro (Admin)
// GET /api/usuarios/admin/usuarios
const obtenerTodosLosUsuarios = async (filtro?: string, page: number = 0, size: number = 10, sort?: string): Promise<ApiResponse<Page<Usuario>>> => {
    return api.get<Page<Usuario>>(`${BASE_PATH}/admin/usuarios`, {
        params: {
            ...(filtro && { filtro }),
            page,
            size,
            ...(sort && { sort })
        }
    });
};

const obtenerUsuarioPorId = async (id: number): Promise<ApiResponse<Usuario>> => {
    return api.get<Usuario>(`${BASE_PATH}/admin/usuario/${id}`);
}

// Solicitar retiro de fondos
// POST /api/usuarios/solicitar-retiro
const solicitarRetiroFondos = async (
    walletAddressId: number,
    monto: number,
    tipoSolicitud: string
): Promise<ApiResponse<string>> => {
    return api.post<string>(`${BASE_PATH}/solicitar-retiro`, null, {
        params: {
            walletAddressId,
            monto,
            tipoSolicitud
        }
    });
};

// Comprar licencia delegada (para otro usuario)
// POST /api/usuarios/comprar-licencia-delegada
const comprarLicenciaDelegada = async (
    tipoLicencia: string,
    destinatario: string,
    tipoMetodoPago: string
): Promise<ApiResponse<string>> => {
    return api.post<string>(`${BASE_PATH}/comprar-licencia-delegada`, null, {
        params: {
            tipoLicencia,
            destinatario,
            tipoMetodoPago
        }
    });
};

// Transferencia entre usuarios
// POST /api/usuarios/transferencia
const transferenciaEntreUsuarios = async (
    usuarioDestinatario: string,
    monto: number,
    tipoWallet: TipoWallets
): Promise<ApiResponse<string>> => {
    return api.post<string>(`${BASE_PATH}/transferencia`, null, {
        params: {
            usuarioDestinatario,
            monto,
            tipoWallet
        }
    });
};

// Aprobar compra de licencia (Admin)
// PUT /api/usuarios/admin/aprobar-licencia/{idSolicitud}
const aprobarCompraLicencia = async (idSolicitud: number): Promise<ApiResponse<string>> => {
    return api.put<string>(`${BASE_PATH}/admin/aprobar-licencia/${idSolicitud}`);
};

// Rechazar solicitud (Admin)
// PUT /api/usuarios/admin/rechazar-solicitud/{idSolicitud}
const rechazarSolicitud = async (idSolicitud: number): Promise<ApiResponse<string>> => {
    return api.put<string>(`${BASE_PATH}/admin/rechazar-solicitud/${idSolicitud}`);
};

// Objeto con todas las funciones
export const usuarioService = {
    editarPerfil,
    verificacionDosPasos,
    obtenerUsuariosEnRed,
    editarUsuarioAdmin,
    solicitarCompraLicencia,
    obtenerSolicitudes,
    obtenerTodosLosUsuarios,
    obtenerUsuarioPorId,
    solicitarRetiroFondos,
    comprarLicenciaDelegada,
    transferenciaEntreUsuarios,
    aprobarCompraLicencia,
    rechazarSolicitud,
};

File: backend/src/main/java/com/api/payglobal/service/usuario/UsuarioService.java
package com.api.payglobal.service.usuario;

import java.math.BigDecimal;
import java.util.List;

import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.userdetails.UserDetailsService;

import com.api.payglobal.dto.request.CambiarPasswordRequest;
import com.api.payglobal.dto.request.EditarPerfilRequest;
import com.api.payglobal.dto.request.GuardarFile;
import com.api.payglobal.dto.request.LoginRequest;
import com.api.payglobal.dto.request.RegistroResquestDTO;
import com.api.payglobal.dto.response.JwtResponse;
import com.api.payglobal.dto.response.SolicitudRetiroDTO;
import com.api.payglobal.dto.response.UsuarioEnRedResponse;
import com.api.payglobal.dto.response.UsuarioExplorerResponseDTO;
import com.api.payglobal.entity.Solicitud;
import com.api.payglobal.entity.Usuario;
import com.api.payglobal.entity.enums.TipoCrypto;
import com.api.payglobal.entity.enums.TipoLicencia;
import com.api.payglobal.entity.enums.TipoMetodoPago;
import com.api.payglobal.entity.enums.TipoSolicitud;
import com.api.payglobal.entity.enums.TipoWallets;

public interface UsuarioService extends UserDetailsService {
        JwtResponse registrar(RegistroResquestDTO registroRequest) throws Exception;

        JwtResponse login(LoginRequest loginRequest) throws Exception;

        Usuario editarPerfilUsuario(EditarPerfilRequest editarPerfilRequest, Long idUsuario) throws Exception;

        void cambiarPasswordUsuario(CambiarPasswordRequest cambiarPasswordRequest, Long idUsuario) throws Exception;

        void cambiarPasswordAdmin(Long idUsuario, String nuevoPassword) throws Exception;

        void verificacionDosPasos(String codigoVerificacion, Long idUsuario) throws Exception;

        List<UsuarioEnRedResponse> obtenerUsuariosEnRed(String username) throws Exception;

        void editarUsuario(Usuario usuario) throws Exception;

        void solicitarCompraLicencia(TipoCrypto tipoCrypto, TipoLicencia tipoLicencia, TipoSolicitud tipoSolicitud,
                        Long idUsuario) throws Exception;

        void solicitarRetiroFondos(Long walletAddressId, BigDecimal monto, TipoSolicitud tipoSolicitud, Long idUsuario)
                        throws Exception;

        void comprarLicenciaDelegada(TipoLicencia tipoLicencia, String destinatario, TipoMetodoPago tipoMetodoPago,
                        Long idUsuario) throws Exception;

        void TransferenciaEntreUsuarios(String usuarioDestinatario, BigDecimal monto, TipoWallets tipoWallet,
                        Long idUsuario) throws Exception;

        void aprobarCompraLicencia(Long idSolicitud) throws Exception;

        void rechazarSolcitud(Long idSolicitud) throws Exception;

        Page<Solicitud> obtenerSolicitudes(Pageable pageable) throws Exception;

        Page<Solicitud> obtenerSolicitudesPorTipos(List<TipoSolicitud> tipos, Pageable pageable) throws Exception;

        Page<Solicitud> obtenerSolicitudesPorUsuario(Long usuarioId, Pageable pageable) throws Exception;

        Page<SolicitudRetiroDTO> obtenerSolicitudesRetiro(Pageable pageable) throws Exception;

        Page<UsuarioExplorerResponseDTO> obtenerTodosLosUsuarios(String filtro, Pageable pageable) throws Exception;

        Usuario obtenerUsuarioPorId(Long idUsuario) throws Exception;

        void eliminarUsuarioPorId(Long idUsuario) throws Exception;

        void aprobarRetiroFondos(Long idSolicitud) throws Exception;

        void rechazarRetiroFondos(Long idSolicitud) throws Exception;

        Resource subirFotoPerfil(GuardarFile guardarFile, Long idUsuario) throws Exception;

        void guardarClaveSeguridad(String claveSeguridad, Long idUsuario) throws Exception;

        Boolean verificarClaveSeguridad(String claveSeguridad, Long idUsuario) throws Exception;

        void aprobarLicenciaMineria(Long idSolicitud) throws Exception; 
}